app/frontend/src/features/messages/messagelist/InfiniteMessageLoader.tsx

```diff
@@ -0,0 +1,94 @@
+import { Box, makeStyles } from "@material-ui/core";
+import CircularProgress from "@material-ui/core/CircularProgress";
+import classNames from "classnames";
+import React, { ReactNode, useCallback, useLayoutEffect, useRef } from "react";
+
+import useOnVisibleEffect from "../useOnVisibleEffect";
+import { messageElementId } from "./MessageView";
+
+const useStyles = makeStyles((theme) => ({
+  loader: {
+    //px for easier scroll compensation
+    width: "100%",
+    position: "absolute",
+    top: 0,
+    paddingTop: theme.spacing(1),
+    "& > *": {
+      display: "block",
+      marginInline: "auto",
+    },
+  },
+  scroll: {
+    ...theme.shape.scrollBar,
+    position: "relative",
+  },
+}));
+
+interface InfiniteMessageLoaderProps {
+  earliestMessageId?: number;
+  fetchNextPage: () => void;
+  isFetchingNextPage: boolean;
+  hasNextPage: boolean;
+  isError: boolean;
+  className?: string;
+  children: ReactNode;
+}
+
+export default function InfiniteMessageLoader({
+  earliestMessageId,
+  fetchNextPage,
+  isFetchingNextPage,
+  hasNextPage,
+  isError,
+  className,
+  children,
+}: InfiniteMessageLoaderProps) {
+  const classes = useStyles();
+
+  const scrollRef = useRef<HTMLElement>(null);
+  const prevScrollHeight = useRef<number | undefined>(undefined);
+  const prevTopMessageId = useRef<number | null>(null);
+
+  const handleLoadMoreVisible = useCallback(() => {
+    prevScrollHeight.current = scrollRef.current?.scrollHeight;
+    if (earliestMessageId) {
+      prevTopMessageId.current = earliestMessageId;
+    }
+    fetchNextPage();
+  }, [earliestMessageId, fetchNextPage]);
+
+  const { ref: loadMoreRef } = useOnVisibleEffect(null, handleLoadMoreVisible);
+
+  useLayoutEffect(() => {
+    if (isFetchingNextPage) return;
+    const messageEl = document.getElementById(
+      messageElementId(prevTopMessageId.current ?? 0)
+    );
+    messageEl?.scrollIntoView();
+  }, [isFetchingNextPage, loadMoreRef]);
```

Comment:
```suggestion
  }, [isFetchingNextPage]);
```
Since `loadMoreRef` is meant to be stable? Also no lint warnings when removed so I think React knows it.